src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -508,34 +422,40 @@ private bool ReadImplementation()
                     // tracks the state of a batch operation while in state Operation.
                     this.ReaderOperationState = OperationState.None;
 
-                    // Also add a pending ContentId header to the URL resolver now. We ensured above
-                    // that a message has been created for this operation and thus the headers (incl.
-                    // a potential content ID header) have been read.
-                    if (this.contentIdToAddOnNextRead != null)
-                    {
-                        this.payloadUriConverter.AddContentId(this.contentIdToAddOnNextRead);
-                        this.contentIdToAddOnNextRead = null;
-                    }
-
                     // When we are done with an operation, we have to skip ahead to the next part
                     // when Read is called again. Note that if the operation stream was never requested
                     // and the content of the operation has not been read, we'll skip it here.
                     Debug.Assert(this.ReaderOperationState == OperationState.None, "Operation state must be 'None' at the end of the operation.");
                     this.State = this.ReadAtOperationImplementation();
+                    if (this.isInChangeset)
+                    {
+                        this.IncreaseChangesetSize();
+                    }
+                    else
+                    {
+                        this.IncreaseBatchSize();
+                    }
+
                     break;
 
                 case ODataBatchReaderState.ChangesetStart:
                     // When at the start of a changeset, skip ahead to the first operation in the
                     // changeset (or the end boundary of the changeset).
-                    this.VerifyReaderStreamChangesetStartImplementation();
+                    Debug.Assert(!this.isInChangeset, "Should have validated that nested changesets are not allowed.");
                     this.State = this.ReadAtChangesetStartImplementation();
+                    if (this.inputContext.MessageReaderSettings.MaxProtocolVersion <= ODataVersion.V4)

```

Comment:
change "<=" to "<" to exclude V4, sinceboth V4 and V4.01 protocols specifies that contentId's uniqueness scope is the whole request.